src/Integration.Vsix/SonarAnalyzerDeactivationManager.cs

```diff
@@ -0,0 +1,137 @@
+using Microsoft.VisualStudio.LanguageServices;
+using SonarLint.Helpers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using SonarLint.VisualStudio.Integration.SonarAnalyzer;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    public class SonarAnalyzerDeactivationManager : IDisposable
+    {
+        internal /*for testing purposes*/ enum ProjectAnalyzerStatus
+        {
+            NoAnalyzer,
+            SameVersion,
+            DifferentVersion
+        }
+
+        private readonly Workspace workspace;
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+
+        private static readonly AssemblyName AnalyzerAssemblyName =
+            new AssemblyName(typeof(SonarAnalysisContext).Assembly.FullName);
+        internal /*for testing purposes*/ static readonly Version AnalyzerVersion = AnalyzerAssemblyName.Version;
+        internal /*for testing purposes*/ static readonly string AnalyzerName = AnalyzerAssemblyName.Name;
+
+        internal /*for testing purposes*/ SonarAnalyzerDeactivationManager(IServiceProvider serviceProvider, Workspace workspace)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            this.workspace = workspace;
+            this.activeSolutionBoundTracker = serviceProvider.GetMefService<IActiveSolutionBoundTracker>();
+
+            SonarAnalysisContext.ShouldAnalysisBeDisabled =
+                tree => ShouldAnalysisBeDisabledOnTree(tree);
+        }
+
+        public SonarAnalyzerDeactivationManager(IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            var componentModel = (IComponentModel)serviceProvider.GetService(typeof(SComponentModel));
+            this.workspace = componentModel.GetService<VisualStudioWorkspace>();
+            this.activeSolutionBoundTracker = serviceProvider.GetMefService<IActiveSolutionBoundTracker>();
+
+            SonarAnalysisContext.ShouldAnalysisBeDisabled =
+                tree => ShouldAnalysisBeDisabledOnTree(tree);
+        }
+
+        private bool ShouldAnalysisBeDisabledOnTree(SyntaxTree tree)
+        {
+            if (tree == null)
+            {
+                return false;
+            }
+
+            IEnumerable<AnalyzerReference> references = workspace?.CurrentSolution?.GetDocument(tree)?.Project?.AnalyzerReferences;
+            var projectAnalyzerStatus = GetProjectAnalyzerConflictStatus(references);
+
+            return HasConflictingAnalyzerReference(projectAnalyzerStatus) ||
+                GetIsBoundWithoutAnalyzer(projectAnalyzerStatus);
+        }
+
+        internal /*for testing purposes*/ bool GetIsBoundWithoutAnalyzer(ProjectAnalyzerStatus projectAnalyzerStatus)
```

Comment:
Can we lazily initialize activeSolutionBoundTracker  at this point instead in the constructor?
This will deffer loading the types/assembly until its needed.